sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AdaptingIntegerArray.java

```diff
@@ -0,0 +1,193 @@
+ * reaching limits. This assumes counts remain low, to lower memory overhead.
+ *
+ * <p>This class is NOT thread-safe. It is expected to be behind a synchronized incrementer.
+ *
+ * <p>Instances start by attempting to store one-byte per-cell in the integer array. As values grow,
+ * this will automatically instantiate the next-size integer array (byte => short => int => long)
+ * and copy over values into the larger array. This class expects most usage to remain within the
+ * byte boundary (e.g. cell values < 128).
+ *
+ * <p>This class lives in the (very) hot path of metric recording. As such, we do "fun" things, like
+ * switch on markers and assume non-null based on presence of the markers, as such we suppress
+ * NullAway as it can't understand/express this level of guarantee.
+ *
+ * <p>Implementations MUST preserve the following:
+ *
+ * <ul>
+ *   <li>If cellSize == BYTE then byteBacking is not null
+ *   <li>If cellSize == SHORT then shortBacking is not null
+ *   <li>If cellSize == INT then intBacking is not null
+ *   <li>If cellSize == LONG then longBacking is not null
+ * </ul>
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
+ * at any time.
+ */
+public class AdaptingIntegerArray {
+  @Nullable private byte[] byteBacking;
+  @Nullable private short[] shortBacking;
+  @Nullable private int[] intBacking;
+  @Nullable private long[] longBacking;
+
+  /** Possible sizes of backing arrays. */
+  private enum ArrayCellSize {
+    BYTE,
+    SHORT,
+    INT,
+    LONG
+  }
+  /** The current byte size of integer cells in this array. */
+  private ArrayCellSize cellSize;
+
+  /** Construct an adapting integer array of a given size. */
+  public AdaptingIntegerArray(int size) {
+    this.cellSize = ArrayCellSize.BYTE;
+    this.byteBacking = new byte[size];
+  }
+
+  /** Creates deep copy of another adapting integer array. */
+  @SuppressWarnings("NullAway")
+  public AdaptingIntegerArray(AdaptingIntegerArray toCopy) {
+    this.cellSize = toCopy.cellSize;
+    switch (cellSize) {
+      case BYTE:
+        this.byteBacking = Arrays.copyOf(toCopy.byteBacking, toCopy.byteBacking.length);
+        break;
+      case SHORT:
+        this.shortBacking = Arrays.copyOf(toCopy.shortBacking, toCopy.shortBacking.length);
+        break;
+      case INT:
+        this.intBacking = Arrays.copyOf(toCopy.intBacking, toCopy.intBacking.length);
+        break;
+      case LONG:
+        this.longBacking = Arrays.copyOf(toCopy.longBacking, toCopy.longBacking.length);
+        break;
+    }
+  }
+
+  /** Add {@code count} to the value stored at {@code index}. */
+  @SuppressWarnings("NullAway")
+  public void increment(int idx, long count) {
+    // TODO - prevent bad index
+    long result;
+    switch (cellSize) {
+      case BYTE:
+        result = byteBacking[idx] + count;
+        if (result > Byte.MAX_VALUE) {
+          // Resize + add
+          resizeToShort();
```

Comment:
Thinking of an edge case where `result` is, for this case, higher than `Short.MAX_VALUE`, in which case `resizeToShort()` will cause overflow.

Reply:
This could be optimised, but effectively it doesn't write the new value.  It resizes to Short *THEN* attempts to write again where it'd detect the overflow.   We can fix this so it resizes tot he appropriate value straight away, happy to fix that in this PR if you'd like.

Reply:
Ah, yeah I missed the recursive call, so it's correct. If you want to optimise it then go ahead, but I'm not too fussed seeing as it's a super edge case and not that much more CPU.

Reply:
In reality, we're mostly bumping counts by 1, so I'm not sure any changes will make it more optimal outside of preserving integer-size once a specific histogram "bumps".